pkg/kubelet/pluginmanager/cache/desired_state_of_world.go

```diff
@@ -0,0 +1,206 @@
+	detailedStr := fmt.Sprintf("(plugin details: %v)", plugin)
+	return generatePluginMsgDetailed(prefixMsg, suffixMsg, plugin.SocketPath, detailedStr)
+}
+
+// GenerateMsg returns simple and detailed msgs for plugins to register
+// that is user friendly and a detailed msg that can be used in logs.
+// The msg format follows the pattern "<prefixMsg> <plugin details> <suffixMsg>".
+func (plugin *PluginInfo) GenerateMsg(prefixMsg, suffixMsg string) (simpleMsg, detailedMsg string) {
+	detailedStr := fmt.Sprintf("(plugin details: %v)", plugin)
+	return generatePluginMsg(prefixMsg, suffixMsg, plugin.SocketPath, detailedStr)
+}
+
+// GenerateErrorDetailed returns detailed errors for plugins to register
+// that can be used in logs.
+// The msg format follows the pattern "<prefixMsg> <plugin details>: <err> ",
+func (plugin *PluginInfo) GenerateErrorDetailed(prefixMsg string, err error) (detailedErr error) {
+	return fmt.Errorf(plugin.GenerateMsgDetailed(prefixMsg, errSuffix(err)))
+}
+
+// GenerateError returns simple and detailed errors for plugins to register
+// that is user friendly and a detailed error that can be used in logs.
+// The msg format follows the pattern "<prefixMsg> <plugin details>: <err> ".
+func (plugin *PluginInfo) GenerateError(prefixMsg string, err error) (simpleErr, detailedErr error) {
+	simpleMsg, detailedMsg := plugin.GenerateMsg(prefixMsg, errSuffix(err))
+	return fmt.Errorf(simpleMsg), fmt.Errorf(detailedMsg)
+}
+
+// Generates an error string with the format ": <err>" if err exists
+func errSuffix(err error) string {
+	errStr := ""
+	if err != nil {
+		errStr = fmt.Sprintf(": %v", err)
+	}
+	return errStr
+}
+
+func (dsw *desiredStateOfWorld) AddHandler(pluginType string, handler PluginHandler) {
+	dsw.Lock()
+	defer dsw.Unlock()
+
+	dsw.handlers[pluginType] = handler
+}
+
+func (dsw *desiredStateOfWorld) GetHandler(pluginType string) (PluginHandler, bool) {
+	dsw.Lock()
+	defer dsw.Unlock()
+
+	h, ok := dsw.handlers[pluginType]
+	return h, ok
+}
+
+func (dsw *desiredStateOfWorld) PluginExistsWithCorrectInfo(pluginInfo PluginInfo) bool {
+	dsw.RLock()
+	defer dsw.RUnlock()
+	if foundInfo, ok := dsw.socketFileToInfo[pluginInfo.SocketPath]; ok {
+		return reflect.DeepEqual(foundInfo, pluginInfo)
+	}
+	return false
+}
+
+func (dsw *desiredStateOfWorld) AddOrUpdatePlugin(pluginInfo PluginInfo) error {
+	dsw.Lock()
+	defer dsw.Unlock()
+
+	if pluginInfo.Name == "" || pluginInfo.Type == "" || pluginInfo.SocketPath == "" || pluginInfo.Endpoint == "" || len(pluginInfo.SupportedVersions) == 0 {
+		return fmt.Errorf("Missing plugin information")
+	}
+	if foundInfo, ok := dsw.socketFileToInfo[pluginInfo.SocketPath]; ok {
+		if reflect.DeepEqual(foundInfo, pluginInfo) {
+			return nil
+		}
+		klog.V(2).Infof("Plugin (Name %s, Type %s, Path %s) exists in desired state cache, but the plugin information is different; updating plugin info", pluginInfo.Name, pluginInfo.Type, pluginInfo.SocketPath)
```

Comment:
So I guess the idea here is that if a plugin get's upgraded. The socket remains the same. But the information (like "supported version" changes. When that happens, since the object in desired state is different from actual state, reconciler will trigger a Unregister operation for that socket, and once the unregister completes, it will start the register operation with the new plugin info (because reconciler starts unregister before register and never lets more then one operation happen at once for the same socket).

That sounds like the right way to handle it. But I'll think over this some more.

CC @msau42 for FYI

Reply:
Ok this logic makes sense to me. It's very nice way of handling the case where a plugin mutates (changes the versions it supports, for example).

Only follow up here:

1) Capture this in a comment here, so it is clear that is the intention.
2) If it is not super painful, add a test that will do this:
    * Create a socket which returns supported versions `0.3.0`.
    * Verify that Register is called correctly to register it.
    * Replace the socket with a socket that returns supported versions `0.3.0, 1.0.0`
    * Verify that the socket is first unregistered and then reregistered.

You can do this in an E2E test in a follow up PR if you want, just make sure to open a tracking bug for it.

Reply:
@saad-ali After my recent update this will longer be true. If a plugin is upgraded while socket remains the same, the plugin will not be re-registered. I made this change because I needed to move the `GetInfo` logic into `operation_generator.go`. Also, the existing `Start(...)` logic of `plugin_watcher.go` trigger registration/deregistration on file create and delete event, and I'm leaving that logic untouched.

Reply:
We were able to reintroduce this logic via timestamps.